[MRG] Add image file reader class

hackermd opened the pull request:
<!--
Please prefix your PR title with [WIP] for PRs that are in progress and change
to [MRG] when you consider them ready for final review.
-->

#### Describe the changes
Adds support for reading efficiently individual frames of a (multi-frame) image without loading the entire Pixel Data element into memory (see also #534, #1263, #1243).

#### Tasks
- [x] Unit tests added that reproduce the issue or prove feature is working
- [x] Fix or feature added
- [x] Code typed and mypy shows no errors
- [ ] Documentation updated (if relevant)
  - [ ] No warnings during build
  - [ ] Preview link (CircleCI -> Artifacts -> `doc/_build/html/index.html`)
- [x] Unit tests passing and overall coverage the same or better

pep8speaks commented:
Hello @hackermd! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pydicom/filereader.py`](https://github.com/pydicom/pydicom/blob/50449ee69dde4e10265e77e455c26e597c67deac/pydicom/filereader.py):

> [Line 1244:80](https://github.com/pydicom/pydicom/blob/50449ee69dde4e10265e77e455c26e597c67deac/pydicom/filereader.py#L1244): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1288:80](https://github.com/pydicom/pydicom/blob/50449ee69dde4e10265e77e455c26e597c67deac/pydicom/filereader.py#L1288): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1331:80](https://github.com/pydicom/pydicom/blob/50449ee69dde4e10265e77e455c26e597c67deac/pydicom/filereader.py#L1331): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1362:80](https://github.com/pydicom/pydicom/blob/50449ee69dde4e10265e77e455c26e597c67deac/pydicom/filereader.py#L1362): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

darcymason commented on pydicom/filereader.py:
```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+    Examples
+    --------
+    >>> from highdicom.io import ImageFileReader
+    >>> with ImageFileReader('/path/to/file.dcm') as image:
```
Reference to highdicom needs replacing

hackermd replied:
Addressed via 8b581786d87423903b0976e3c1b227273a893a81

darcymason commented on pydicom/filereader.py:
```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+    """
+
+    def __init__(self, filename: str):
+        """
```
Can we also include a file object rather than just a filename?  Inevitably someone needing to work with streams will ask for it, although some may be limited if there are negative `seek`s, as seen for example in #753.

hackermd replied:
I generally the idea of allowing the user to alternatively provide an existing file object and have refactored the class accordingly via e1056bd16e193cfd8495cf2e9f690d9fe03cc5a4.

hackermd replied:
I have decided to not set the `is_little_endian` and `is_implicit_VR` attributes on the file object, so that the object does not get modified. Instead an error is raised if their values are not correct.

This is one of the reasons I initially implemented the class such that the file object gets created under the hood and the attributes are set correctly.

@darcymason What are your thoughts on this?

darcymason commented on pydicom/filereader.py:
```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+
+
+class ImageFileReader(object):
+
```
Subclassing from `object` only needed for Python 2, which we have dropped.

hackermd replied:
Addressed via 0b5b68cd60e5f8e31f9ed9e140a52458faeba32c

darcymason commented on pydicom/filereader.py:
```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+    >>> from highdicom.io import ImageFileReader
+    >>> with ImageFileReader('/path/to/file.dcm') as image:
+    ...     print(image.metadata)
+    ...     for i in range(image.number_of_frames):
```
I'm not a fan of the term `metadata` here, especially because it sounds too much like the File Meta Information defined for part 10 files in the Standard.  But also because it is really a Dataset (from `dcmread` on `open` call) of all data elements except the pixel data; pydicom (and DICOM itself) don't give unique status to pixel data, unlike other file types with a 'header' or 'metadata'; in DICOM they are all just data elements.

Having said that, I'm not sure what to call it.  In pydicom, using `stop_before_pixels=True` just gives a Dataset, simply with the pixel data element missing.  I'll mull this over a little more...

hackermd replied:
In DICOMweb, the term [metadata](http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_10.4.html#sect_10.4.1.1.2) is used to refer to the subset of Data Elements in a Data Set that are not considered [bulkdata](http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_10.4.html#sect_10.4.1.1.5) (i.e., that have a size smaller than a defined threshold).

I generally use to term to indicate that a given `pydicom.dataset.Dataset` may not contain the Pixel Data element.

mrbean-bremen replied:
I have heard that term too, and have also been confused by it. I think it comes from image formats like TIFF who have embedded metadata tags.
What I encounter more often is "DICOM header data" - not sure if that would be a better term.

darcymason replied:
> DICOM header data

I think 'header' applies to other formats where the pixel data is not tagged, but whatever is left after some initial data is just raw pixel bytes.  It assumes every file is basically pixel data, plus a few extra things.

In DICOM, many file types (SOP Classes) do not have pixel data, so the whole file would be 'header', which doesn't really make sense.  Same goes for `metadata` in the sense of everything that is not pixels.

> could we have something like a `frames` property as an alternate to `pixel_array`?

Any thoughts on the above general comment?  If this could work as an add-on to `Dataset`, then the nomenclature issue could go away.

mrbean-bremen replied:
> In DICOM, many file types (SOP Classes) do not have pixel data, so the whole file would be 'header', which doesn't really make sense.

Fair enough. I didn't really like the term too much, was just thinking of an alternative to metadata.

> Any thoughts on the above general comment? 

I think it makes sense. Frames are an inherent property of DICOM files, so it would fit IMHO. Also, I think that the ability to read single frames is an important feature that belongs into the core.

hackermd replied:
Regarding terminology, I think that anyone who uses DICOMweb will be familiar with the term "metadata" and the value returned by the `metadata` property will (more or less) correspond to what one would get via a DICOMweb `/metadata` call. Similarly, the value returned by `read_frame()` corresponds to what one would get via a DICOMweb `/frames/{frameNumber}` call.

darcymason reviewed (commented):
Just a few thoughts from a first pass...

Re my comment on `metadata`, I wonder if there is a way to integrate this into `Dataset` - just off the cuff here, could we have something like a `frames` property as an alternate to `pixel_array`?  Generally I don't like to complicate `Dataset` with new code, but this is quite a common need, and important for large file sizes.

hackermd commented:
> Re my comment on metadata, I wonder if there is a way to integrate this into Dataset - just off the cuff here, could we have something like a frames property as an alternate to pixel_array? Generally I don't like to complicate Dataset with new code, but this is quite a common need, and important for large file sizes.

I initially also thought about adding this functionality to the `Dataset` class. However, I meanwhile came to the conclusion that this would not be desirable for the following reasons:

- Efficient frame-level access would require setting `stop_before_pixels` to `True` when using `pydicom.fileread.dcmread()` to read the data set from a file, which would tightly couple the two abstractions. Otherwise, the entire Pixel Data element would be fully read into memory anyways.
- If the `Dataset` does not get read from a file, but is instead retrieved over network, the whole Pixel Data element will already be available in memory. One could avoid decoding of all frames and only decode an individual frame, but this could be a separate feature to add in the future.
- Not every `Dataset` actually represents an image and I therefore find adding more image-specific methods to the class problematic. We could discuss the implementation of an `Image` class that gets derived from `Dataset`, which would get us very close to the approach we've chosen in *highdicom* and construct the subclass when reading data set from a file. However, this would break backwards compatibility.

hackermd commented:
@darcymason not sure whether you noticed the color correction feature (see [here](https://github.com/hackermd/pydicom/blob/e1056bd16e193cfd8495cf2e9f690d9fe03cc5a4/pydicom/filereader.py#L1592-L1615)), which depends on https://github.com/pydicom/pydicom/pull/1446. This feature is potentially a bit tricky, because it requires Pillow. We could also leave that out for now.

darcymason commented:
> * Efficient frame-level access would require setting `stop_before_pixels` to `True` when using `pydicom.fileread.dcmread()` to read the data set from a file

This had crossed my mind, but I was thinking (without saying it 'out loud' of course) that we might want to move in this direction anyway with memory mapping, as noted in previous discussions like #1267.  I think it might be interesting to explore whether memmap'ng can be used on all files (or at least ones above a modest size) transparently, without impacting performance of reading smaller files.  I think everyone would dealing with large files would like that behavior, and if was transparent in other cases, no one would complain.

I've also was re-reading the discussion about combining `config` items (e.g. from around here https://github.com/pydicom/pydicom/issues/1232#issuecomment-827810424) and thinking about the previous request to make config non-global, and thinking perhaps we can incorporate that into the `dcmread` call.  Very early thinking though.  A more general config system could make it easier to turn memmapping on/off with fine control.

> * If the `Dataset` does not get read from a file, but is instead retrieved over network, the whole Pixel Data element will already be available in memory.

I'm not clear on how the separate class is different in this respect.


> * Not every `Dataset` actually represents an image and I therefore find adding more image-specific methods to the class problematic.

I generally agree very much with avoiding image-specific methods, but we already have `pixel_array` which breaks this distinction, so to me adding a `frames` property is just an evolution which may be worth the complication.

I'll keep mulling all this though, happy to hear other thoughts...

hackermd commented:
> > Not every Dataset actually represents an image and I therefore find adding more image-specific methods to the class problematic.

> I generally agree very much with avoiding image-specific methods, but we already have pixel_array which breaks this distinction, so to me adding a frames property is just an evolution which may be worth the complication.

I would consider implementing a `get_frames()` method rather than a `frames` property. This will allow parameterization if this should become necessary (for example, to correct color of SM image frames or HU values of CT image frames).

hackermd commented:
> > Efficient frame-level access would require setting stop_before_pixels to True when using pydicom.fileread.dcmread() to read the data set from a file

> This had crossed my mind, but I was thinking (without saying it 'out loud' of course) that we might want to move in this direction anyway with memory mapping, as noted in previous discussions like #1267. I think it might be interesting to explore whether memmap'ng can be used on all files (or at least ones above a modest size) transparently, without impacting performance of reading smaller files. I think everyone dealing with large files would like that behavior, and if was transparent in other cases, no one would complain.

Personally, I would distinguish between the representation of a DICOM Data Set in memory and a DICOM Part 10 file stored on disk. We should not assume that every `pydicom.dataset.Dataset` was read from a file on disk or that the file still exists after the data set was (partially) read. For example, the [dicomweb-client](https://github.com/MGHComputationalPathology/dicomweb-client) retrieves data sets from a server via DICOMweb. I tend argue that the `pydicom.dataset.Dataset` class should not deal with file I/O, but this should be left to separate file reader classes/functions.

However, independent of whether we'll further pursue the memory mapping, the `ImageFileReader` class would be nice to have in my opinion. The class hides all the implementation details for reading and decoding individual frames, but those could be factored out and reused for other purposes.

darcymason commented:
> would consider implementing a `get_frames()` method rather than a `frames` property. This will allow parameterization if this should become necessary

Good point.

> I tend argue that the `pydicom.dataset.Dataset` class should not deal with file I/O, but this should be left to separate file reader classes/functions.
> 
> However, independent of whether we'll further pursue the memory mapping, the `ImageFileReader` class would be nice to have in my opinion. The class hides all the implementation details for reading and decoding individual frames, but those could be factored out and reused for other purposes.

That seems reasonable - we could start here and perhaps incorporate into Dataset at some point in future, but with the logic kept in the separate class.

hackermd commented:
@darcymason I removed the color management code from the `ImageFileReader` for now, since it depends on #1446, which is probably more WIP than MRG at this point (I updated the PR title to reflect that sentiment).

We will need Pillow for color management (at least in its current implementation). However, Pillow is an optional dependency and therefore requires (not so elegant) workarounds to avoid Pillow symbols from being exposed globally. I would suggest working on getting the `ImageFileReader` into the library without color correction. We can revisit this later should #1446 get merged. However, the correction could (and maybe even should) be done separately outside of the `ImageFileReader`.

hackermd commented:
@darcymason are there any outstanding TODOs preventing this PR from being merged and released?

darcymason commented:
> @darcymason are there any outstanding TODOs preventing this PR from being merged and released?

Not from me, but I was hoping @scaramallion could review and comment on how/whether this fits in with the various handlers concepts.

hackermd commented:
@scaramallion what are thoughts?

hackermd commented:
@scaramallion I was wondering whether you had a chance to take a look. Please let me know if you have any questions or comments. It would be great if we could get clarity on whether we can include this functionality sometime soon.

hackermd commented:
@darcymason @scaramallion is this something you still think would be useful to add to pydicom?

mrbean-bremen commented:
> is this something you still think would be useful to add to pydicom?

In my opinion, this would be very useful. I just stumbled over this PR because of a related SO question and I'm not sure in what state it is, but I had to access single frames of large multi-frame in the past and find it an important feature (we had helped dcmtk to add a similar feature at the time).
@scaramallion - can be revive this, or have you been working on a similar feature? I dimly remember some related discussion, but could not find it...

scaramallion commented:
Closing, implemented in `pixels` module
